### PR TITLE
feat: add sound effects

### PR DESCRIPTION
**Pull request by QuakeEye**
## Description

This pull request adds sound effects to the player for walking and weapons (both melee and ranged).

## Testability

Open the scene for this ticket (73), then walk around and use a weapon, you can also test the ranged weapon sound effects by using a crossbow instead of a melee weapon.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

File: aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs [hunk 1]
```diff
@@ -109,6 +120,13 @@ private void MovePlayer()
 
         if (isWallColliding) HandleWallCollision(wallNormal);
         LimitVelocity(isOnSlope);
+
+        // Lastly, play a footstep sound when the player is grounded, at certain intervals
+        if (_isGrounded && _rigidbody.velocity.magnitude > 0.25f && _timer.Countdown <= 0)
```
Comment (SilasPeters) [inline aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs:125]: The Timer class should have a method `Passed` or `TimeUp` to check this. This reads more intuitively than `.Counterdown <= 0`.

File: aplib.net-demo/Assets/Scripts/Entities/Player/Weapons/MeleeWeapon.cs [hunk 2]
```diff
@@ -28,10 +28,11 @@ public class MeleeWeapon : Weapon
     /// Ensure the height is at least twice the radius, because the height of the hitzone (capsule) must at least be the diameter of the spheres.
     /// (If the two spheres are at the same position, the capsule just becomes a sphere and height = 2 * radius)
     /// </summary>
-    private void Start()
+    private new void Start()
```
Comment (SilasPeters) [inline aplib.net-demo/Assets/Scripts/Entities/Player/Weapons/MeleeWeapon.cs:31]: Overrride the new with override

**Review by SilasPeters — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Util/Timer.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+using UnityEngine;
+
+public class Timer : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scripts/Util/Timer.cs:3]: Not spending too much time on the Timer as it won't even be used for sound effects in about next sprint

File: aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs [hunk 4]
```diff
@@ -34,13 +38,20 @@ public class Movement : MonoBehaviour
...
 
+    private EntitySound _footStep;
+    private Timer _timer;
+
     /// <summary>
     /// Initialize the player's rigidbody and collider, and freeze the player's rotation.
     /// </summary>
     private void Start()
     {
+        // Link the components of this object to the variables
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs:49]: Why not put this in the summary?

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs [hunk 5]
```diff
@@ -33,7 +33,7 @@ public class MeleeWeapon : Weapon
     /// Ensure the height is at least twice the radius, because the height of the hitzone (capsule) must at least be the diameter of the spheres.
     /// (If the two spheres are at the same position, the capsule just becomes a sphere and height = 2 * radius)
     /// </summary>
-    private void Awake()
+    void Start()
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:36]: Why did you change this to void Start with no access modifier (probably accidental)?

**Review by JoenvandeVorle — CHANGES_REQUESTED** [review]
Some of the tests are failing when I run them (melee and first aplib test)

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs (hunk 5 above)
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:36]: this used to override a start method, private is not necessary as the default is private

File: aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs (hunk 4 above)
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs:49]: Because it is a comment on that block of code and not the entire function, wanted to group it and then comment it for clarity.

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs (hunk 5 above)
Comment (JoenvandeVorle) [reply aplib.net-demo/Assets/Scripts/Entities/Weapons/MeleeWeapon.cs:36]: The rest of the code (even your own) adds private keywords to `Awake `and `Start `though, so I would still add it for consistency.

File: aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+using UnityEngine;
+
+[RequireComponent(typeof(AudioSource))]
+public class EntitySound : MonoBehaviour
+{
+    [SerializeField] private AudioClip[] _footSteps;
+    [SerializeField] private AudioClip[] _knifeSwings;
+    [SerializeField] private AudioClip[] _crossbowShots;
+    [SerializeField] private AudioClip[] _deathSounds;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs:9]:
```suggestion
    [SerializeField]
    private AudioClip[] _footSteps;
    
    [SerializeField]
    private AudioClip[] _knifeSwings;
    
    [SerializeField]
    private AudioClip[] _crossbowShots;
    
    [SerializeField]
    private AudioClip[] _deathSounds;
```

Comment (JensSteenmetz) [reply aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs:9]: There is a resharper setting for this in the editorconfig, consider using resharper.

File: aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    private AudioSource _audioSource;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/EntitySound.cs:10]: Nit
```suggestion

    private AudioSource _audioSource;
```

File: aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs [hunk 8]
```diff
@@ -80,6 +84,7 @@ public void Reset()
     /// </summary>
     private void Die()
     {
+        _entitySound.Die();
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs:87]: Definitely don't do this. Some entities with Health don't need sounds. This is what the Death event is for. You can subscribe the sound to it from another script, for example the EntitySound script.
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs [hunk 9]
```diff
@@ -41,9 +42,12 @@ public class HealthComponent : MonoBehaviour
     [SerializeField]
     private int _maxHealth = 100;
 
+    [SerializeField] private EntitySound _entitySound;
+
     private void Awake()
     {
         Health = _maxHealth;
+        _entitySound = GetComponent<EntitySound>();
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs:50]:
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs [hunk 10]
```diff
@@ -41,9 +42,12 @@ public class HealthComponent : MonoBehaviour
     [SerializeField]
     private int _maxHealth = 100;
 
+    [SerializeField] private EntitySound _entitySound;
+
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs:46]:
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs [hunk 11]
```diff
@@ -1,6 +1,7 @@
 using System;
 using UnityEngine;
 
+[RequireComponent(typeof(EntitySound))]
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/HealthComponent.cs:4]:
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs [hunk 12]
```diff
@@ -1,5 +1,9 @@
 using UnityEngine;
 
+[RequireComponent(typeof(Rigidbody))]
+[RequireComponent(typeof(CapsuleCollider))]
+[RequireComponent(typeof(EntitySound))]
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs:5]: Don't think you want to do sound stuff here either. Though currently we don't have a Move event.

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs [hunk 13]
```diff
@@ -34,5 +34,8 @@ public override void UseWeapon()
             HealthComponent enemy = hit.collider.GetComponent<HealthComponent>();
             enemy?.ReduceHealth(_damage);
         }
+
+        // Play a random whoosh crossbow sound.
+        _entitySound.Shoot();
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/RangedWeapon.cs:39]: Again, don't do this here.

**Review by JoenvandeVorle — APPROVED** [review]

File: aplib.net-demo/Assets/Scripts/Entities/Weapons/Weapon.cs [hunk 14]
```diff
@@ -5,6 +5,9 @@
 /// </summary>
 public abstract class Weapon : MonoBehaviour
 {
+    // The player sound component used to play weapon sounds.
+    [SerializeField] protected EntitySound _entitySound;
+
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Entities/Weapons/Weapon.cs:10]: Not here

File: aplib.net-demo/Assets/Scripts/Util/Timer.cs [hunk 15]
```diff
@@ -1,34 +1,30 @@
...
-/// <summary>
-/// Basic timer class 
-/// </summary>
-public class Timer
+public class Timer : MonoBehaviour
 {
-    private float _time;
-    private readonly float _duration;
+    [SerializeField] private float _minTime = 0.1f;
+    [SerializeField] private float _maxTime = 0.15f;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Util/Timer.cs:6]:
```suggestion
    [SerializeField]
    private float _minTime = 0.1f;
    [SerializeField]
    private float _maxTime = 0.15f;
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Looks good overall, left some comments on separation of sound system.

File: aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs (hunk 12 above)
Comment (QuakeEye) [reply aplib.net-demo/Assets/Scripts/Entities/Player/Movement.cs:5]: This will be replaced when animation clips get added, as discussed irl

**Review by JensSteenmetz — APPROVED** [review]
LDTMFN (Look Decent To Me For Now)

**Review by JoenvandeVorle — APPROVED** [review]
